vowpalwabbit/cb_explore_adf_cover.cc

```diff
@@ -160,6 +177,17 @@ void cb_explore_adf_cover::predict_or_learn_impl(LEARNER::multi_learner& base, m
   sort_action_probs(_action_probs, _scores);
   for (size_t i = 0; i < num_actions; i++) preds[i] = _action_probs[i];
 
+  if (VW_DEBUG_LOG)
```

Comment:
Normal compiler optimization should do the same thing without contexpr.  Are there other reasons besides that?  For example see https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tQVvQAZPLUwA5YwCNMxEJIDspAA6oFhdbT1DE0FvXzU6Kxt7IycXd0VlTFV/BgJmYgJA41NOBJVw2lT0gki7R2dXDwU0jKzg3Ori0ujYyoBKRVQDYmQOAHIpAGZrZEMsAGpxQZ11auJMZiMp7HEABgBBNfW5g1VxgDcAdwB9dAdgLfE3WQ3x8bRaavHGtWRxh1R9cawqZgNWAjHVioYB3MFTAAi41%2BrCUUxu6zuDyeLzwbw%2BX3Y6RsxDBeLukOhbDhgwRSLoKLSr3en1YzwWPQQ%2BPxhJhJLJ9wpBGeVLRNK%2BwHQzOFk0GULZmHhW3Jj25qPRtPGguOnnm%2BD2LLFRNhktJ0s5sp5zGpGLpDms6QAniKCVqJVLbgbKca%2Bab7g5jsx0FQbaLxcTdRzkXLeQqvsgFKhmMxfX7tez9cGjSbFRHPcl9phhayAw7No6k/L%2BXSCMRmMgANbHRqVuM/P4AoEgyZuHTjLHEHEtttKdLIJlXNuC7tK9AqtVo7mD94W4jW6fID1en0LyPRkdp8tqTN5q4Q3cbIY/axZgBqAHVjhDsNIAKoAcWOFgA8veDiczqAQPX/oDgRdDw2aw5QARwMdJMAgNoWw5O48CoCAjlOc4QBADscTafVmXmAhulocZJE4eFJkdOCEKQz9UNLcsqxrCtMNI/EcLwgjJGIrDxngiALyvG8HyfV8GMRYVmM7AjBnYwDhLuUT8O4CS9UPNwIT6DpWBAPoAFY%2BlIUw%2BlWHTUA0nQ5DkZ4uh6LMhk4HSCA0gy2g6CsQE01YhA07gdL0gzSCMvodIUEA3Ls/TVNIOBYCQNAjE8PB2DICgIGi2L4pAYAAE5JFIKg4oIZxAogBx7J081aCtDSbNIaKjC0Ahn1oVhLWK0gsCMERgHYZr8HmDNMEC0LSEwAAPJIDDyiqdOA5RmtYPAHDLOc9CwCbSFLPAjAmjoaHoJg2A4Hh%2BEEYRRBQUyZCEObAsgDpUE8Ap%2BoAWmfSQAsSZINAgcw6hyQbtGacoXFyUI/Dob6Qh8EHaH%2BmIKgaN6CiKWp9GyQQlHyFIahKE8WlhxRMbBhpMeh1pOA6SNul6Lg1I07TdOavyhoADgANge5nuCVZA3nSgA6SRxggXBCBISZJEGXJxj0GK4ucUXxegkyZDkWzisc0hnNc9y%2Bk8umBr8gKgtW1Xqb6F7dZ8/WjdCtXM2IXwNG4IA

Reply:
They are logically compile time available constants and so should be constexpr. The compiler can deal with them and optimize it which is good but if we wanted to use them in some larger constexpr expression it would not be possible without constexpr.

Reply:
Done.  FYI...  Under msvc 2019 this still does not get compiled out unless -O is specified.  gcc seems to do the right thing.